### ISSUE TITLE
ActiveRecord.where(hash).create callbacks inherit the where(hash)

### ISSUE THREAD

**Issue by brocktimus** [rails/rails#9894]
Running rails 4.0.0-beta1 on ruby 2.0.0-p1.

It seems as though when you perform code like the below, the where clauses are inherited by any callbacks it runs.

``` ruby
Category.where(full_name: 'root/bar').create
```

The above exhibits the problem. All of the below function as intended.

``` ruby
Category.new(full_name: 'root/bar').save
Category.create(full_name: 'root/bar')
Category.where(full_name: 'root/bar').new.save
```

By debugging on the rails console it reveals:

``` ruby
# This is basically running
Category.where(full_name: 'root/bar').where(full_name: 'root').first

# When this is expected
Category.where(full_name: 'root').first
```

This also applies to the first_or_create methods. I'm yet to test the find_or_create methods or the rails 3.x series. Should I write this into an active_record test? Any advice as to where to start?

I've created an example application with this problem here https://github.com/brocktimus/rails_create_differences. Example class used below.

``` ruby
class Category < ActiveRecord::Base

  has_many :children, class_name: 'Category', foreign_key: :parent_id
  belongs_to :parent, class_name: 'Category'

  before_validation :assign_parents, on: :create

  private

  def assign_parents
    parts = full_name.split '/'
    self.name = parts.pop
    self.parent = Category.where(full_name: parts.join('/')).first_or_initialize unless parts.empty?
  end

end
```

**Comment by jarmo** [comment]
The problem still exists in `rails 4.2.0` - i can reproduce the problem with example described in #12305.

**Comment by ThomasAlxDmy** [comment]
@rafaelfranca Do you realize the bug has been open for 2 years and it's still not fixed? WOW...

**Comment by brocktimus** [comment]
@ThomasAlxDmy its OSS, feel free to open a PR with a fix. Comments like that really aren't helpful.

@rafaelfranca @senny since this has been behaviour for all of the 4.x series is it worth reconsidering what intended behaviour should be given 5 is on the horizon? Possibly more people relying on this behaviour now than not. I'm not sure.

**Comment by ThomasAlxDmy** [comment]
@brocktimus Sorry don't want my name associated with rails ;)

**Comment by dhh** [comment]
Too late. Your entitlement to the free labor of others is now a permanent marker in the annals of Rails.

**Comment by ThomasAlxDmy** [comment]
@dhh lmao When did I say I was using rails? Was just trying to help a friend, I dropped rails/AR years ago mostly because of that kind of issues ;)

**Comment by dhh** [comment]
Your friend is most welcome to join the community and help device a solution to their issue, if they can do so without your obnoxious entitlement. 

Please do be on your merry way to whatever community you've found that has either no issues or is willing to tolerate your attitude about them.

**Comment by ThomasAlxDmy** [comment]
@dhh You're over reacting my friend, I'm just pointing out true facts. and that's it. There ain't no bad intention. We all have issues but there's way different ways to tackle them. I was reading an interesting article about why is rails community slowly dying (small comparison with node -- which I'm not fan of personally) http://www.google.com/trends/explore?hl=en-US#q=%22ruby+on+rails%22,+/m/0bbxf89&cmpt=q&tz&tz and I do believe that is that kind of attitude that leads to it. 

I'm sorry if you can't distinguish your allies from your enemies.

**Comment by dhh** [comment]
Your "true facts" are as laughable as your claim of "ally". Rails is at this point a massive system with an incredibly strong, diligent, and ever-expanding community of users and contributors working on improving it. 

But there will always be bugs, and this particular bug just wasn't important enough to summon any effort to its fix. Ipso facto. 

Rails is not a vendor that owes you, or your friend, anything. It's a common space where everyone can contribute their own fixes and improvements, and in return enjoy that in kind from others.

You contributed nothing but needless scorn and entitlement. That is not the behavior of an "ally". Your post-rationalization for this disgraceful behavior is that the truth of a fact like "bug X has not been fixed for me in duration Y" completely side-steps the point. That truth is both trivial and obvious. You're not providing any service of revelation by sharing.

The tone in which you chose to, though, was indeed in bad intention. Your clearly meant to, and obviously continue to, belittle the work of thousands of contributors who share their work with all for free. 

And for what? Do you think anyone will be inspired to help your friend deal with this issue with greater haste because you huffed and puffed? Quite the contrary. Demanding the free labor of others with indignation is no way to inspire their charity.

So, again, please pack your entitlement, your banal analysis of the health of Rails, and be gone. Shoo, shoo.

**Comment by rafaelfranca** [comment]
Since we are discussing tends, these are interesting number about what you are complaining here http://issuestats.com/github/rails/rails. I think these numbers speak by their self.

Yes, I'm aware that this issue is open for 2 years. In fact I'm aware of all the issues on the issue tracker. But my time is limited, my interest too and I try to focus my time on the things that most interest me, because after all I work here not because someone pays me, but because I love to work with this team and use my free time to help people.

I'm also aware that this is not a trivial issue to fix and it will require a lot of work and it may break a lot of applications like @brocktimus correctly pointed. I know It will be fixed eventually, because there are so many great people working on this framework that it is just matter of time. Every single day someone new starts to contribute to this framework and it is amazing how a little bit of encouragement can create amazing regular contributors like @senny @chancancode @sgrif @kaspth @robin850 @vipulnsward @meinac just to name few.

I really don't know if Rails is dying, but I don't care, because the machinery that runs this framework is more alive than ever.

**Comment by rafaelfranca** [comment]
@brocktimus indeed, this is something that we can just change on Rails 5. The issue is assigned to me, we already have some patches, @sgrif already touched this code too, so I believe we will get it fixed in time to Rails 5.

**Comment by ThomasAlxDmy** [comment]
@dhh Actually I don't think you're right. Because now I got your full attention, then people will come watch to see what's happening, and that issue will be work on again. Awareness is the key.

`Do you realize the bug has been open for 2 years and it's still not fixed?`

IS a true fact, and if you want to laugh about it, feel free but then I really wonder what kind of leader you are for "Your community". As you see it was not address against anyone, it was just an open statement. I have not asked for anything, nor I said that you owe me anything. 

Do you think you're going to win that argument by insulting and trolling me? In my humble opinion you should focus on more important thing for "Your community" than attacking the ones that want to raise awareness. 

For my friends we found a way around it, so now it would just be helping "Your community"... Again, I think your attitude is very disappointing far far away from constructive. You should learn that contributing to a project is not the only way to make it move forward (note that I do not claim I ever did), and you should be grateful to all the people that are still trying to make Rails a better word rather than trolling them for your own sake.

**Comment by rafaelfranca** [comment]
> @dhh Actually I don't think you're right. Because now I got your full attention, then people will come watch to see what's happening, and that issue will be work on again. Awareness is the key.

In my experience that is not what happens. In fact this issue is an outlier, for the reasons that I already explained. So even with awareness the fix is still not trivial and I doubt it will be fixed.

**Comment by jarmo** [comment]
@ThomasAlxDmy I love statistics. Check this out https://www.google.com/trends/explore?hl=en-US#q=rails%2C%20%2Fm%2F0bbxf89&cmpt=q&tz=Etc%2FGMT-3

**Comment by joemsak** [comment]
You left out the `WOW...` part of your quote, dude, which was kind of the finer end of your shitty point

**Comment by ThomasAlxDmy** [comment]
@joemsak stay polite please ;). And Yes WOW to express how surprise I was, did not bring much in the conversation. Instead of trolling feel free to leave constructive comment and show that you're a grown man.

To come back to what Rafael said, in my humble opinion having bug is one things, not documenting and letting them hide into the code is another one.

**Comment by joemsak** [comment]
You've made an excellent point about leaving constructive comments.

**Comment by rafaelfranca** [comment]
If we document a bug it is not a bug anymore but a feature :smile:. Like I said this bug will die, when we think it is time.

**Comment by ThomasAlxDmy** [comment]
😂😂 That's an extreme way of seeing things but why not...

**Comment by CaseyLeask** [comment]
Please leave these wonderful people alone. Myself and many others appreciate their work.

**Comment by shakycode** [comment]
Seriously, this guy has become a troll himself. The issue will be fixed when it's fixed. The Rails core team along with the entire community is constantly working on both features and bugs which amazes me.  Personally, I'm grateful for all of their hard work and sacrifice for the betterment of the framework. 

If people want to bitch and whine about a long standing "bug" then perhaps they should quit the rant, write some code, and create a PR.  Last time I checked the core team and contributors are not on some corporate payroll and are not obligated to do anything. They do it for the love of the framework and the community which I think is fabulous. 

Hopefully this thread dies, it's getting way out of hand. 

Myself and thousands others appreciate all you guys do!

**Comment by prusswan** [comment]
Certainly, this is just one of many issues with AR (the other one which comes to mind being rails/rails#9813 which can be "solved" by upgrading, which is not always a practical solution)

I don't even recall the reason why I subscribed to this, but over time I learnt to be more defensive against the very framework itself (especially functionality that is obscure or only used by a minority).

**Comment by brocktimus** [comment]
@rafaelfranca good to hear. Doesn't worry me too much since I just found it when playing with pre release stuff.

@prusswan if ever I'm like "I wonder what happens when I..." and I think the behaviour is a bit edge case-y I'll just write a test around the "broader" task I'm trying to accomplish. Caught a few bugs in similar cases where I've been relying upon edge cases which have changed during an upgrade.

**Comment by brocktimus** [comment]
@sgrif based on your comment in https://github.com/rails/rails/issues/23286#issuecomment-178274217 should this issue also be closed since its currently targeting 5.0?
